// eslint-plugin-playwright/lib/rules/no-skipped-test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ast_1 = require("../utils/ast");
exports.default = {
    create(context) {
        return {
            CallExpression(node) {
                const options = context.options[0] || {};
                const allowConditional = !!options.allowConditional;
                const { callee } = node;
                if (((0, ast_1.isTestIdentifier)(callee) || (0, ast_1.isDescribeCall)(node)) &&
                    callee.type === 'MemberExpression' &&
                    (0, ast_1.isPropertyAccessor)(callee, 'skip')) {
                    const isHook = (0, ast_1.isTestCall)(node) || (0, ast_1.isDescribeCall)(node);
                    // If allowConditional is enabled and it's not a test/describe hook,
                    // we ignore any `test.skip` calls that have no arguments.
                    if (!isHook && allowConditional && node.arguments.length) {
                        return;
                    }
                    context.report({
                        messageId: 'noSkippedTest',
                        node: isHook ? callee.property : node,
                        suggest: [
                            {
                                fix: (fixer) => {
                                    return isHook
                                        ? fixer.removeRange([
                                            callee.property.range[0] - 1,
                                            callee.range[1],
                                        ])
                                        : fixer.remove(node.parent);
                                },
                                messageId: 'removeSkippedTestAnnotation',
                            },
                        ],
                    });
                }
            },
        };
    },
    meta: {
        docs: {
            category: 'Best Practices',
            description: 'Prevent usage of the `.skip()` skip test annotation.',
            recommended: true,
            url: 'https://github.com/playwright-community/eslint-plugin-playwright/tree/main/docs/rules/no-skipped-test.md',
        },
        hasSuggestions: true,
        messages: {
            noSkippedTest: 'Unexpected use of the `.skip()` annotation.',
            removeSkippedTestAnnotation: 'Remove the `.skip()` annotation.',
        },
        schema: [
            {
                additionalProperties: false,
                properties: {
                    allowConditional: {
                        default: false,
                        type: 'boolean',
                    },
                },
                type: 'object',
            },
        ],
        type: 'suggestion',
    },
};
